grpc/src/test/java/io/stargate/grpc/service/MessageStreamObserverTest.java

```diff
@@ -0,0 +1,119 @@
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.BiFunction;
+import org.junit.jupiter.api.Test;
+import org.mockito.ArgumentCaptor;
+
+class MessageStreamObserverTest {
+  @Test
+  @SuppressWarnings("unchecked")
+  public void shouldOnCompleteWhileThereIsNoInFlight() {
+    // given
+    StreamObserver<QueryOuterClass.Response> callerStreamObserver = mock(StreamObserver.class);
+    StreamObserver<QueryOuterClass.Response> responseStreamObserver =
+        new SynchronizedStreamObserver<>(callerStreamObserver);
+    ExceptionHandler exceptionHandler = new ExceptionHandler(responseStreamObserver);
+    BiFunction<QueryOuterClass.Response, AtomicLong, MessageHandler<QueryOuterClass.Response, ?>>
+        handlerProducer =
+            (response, inFlight) -> {
+              inFlight.decrementAndGet(); // inFlight is decremented
+              return mock(MessageHandler.class);
+            };
+    ScheduledExecutorService executor = Executors.newScheduledThreadPool(1);
+
+    MessageStreamObserver<QueryOuterClass.Response> observer =
+        new MessageStreamObserver<>(
+            responseStreamObserver, exceptionHandler, handlerProducer, executor, 10);
+
+    // when
+    observer.onNext(QueryOuterClass.Response.newBuilder().build());
+
+    // then should immediately complete and call the caller onComplete
+    assertDoesNotThrow(observer::onCompleted);
+    verify(callerStreamObserver, times(1)).onCompleted();
+  }
+
+  @Test
+  @SuppressWarnings("unchecked")
+  public void shouldWaitWithOnCompleteWhileThereAreRequestsInFlight() {
+    // given
+    ScheduledExecutorService executor = Executors.newScheduledThreadPool(2);
+    StreamObserver<QueryOuterClass.Response> callerStreamObserver = mock(StreamObserver.class);
+    StreamObserver<QueryOuterClass.Response> responseStreamObserver =
+        new SynchronizedStreamObserver<>(callerStreamObserver);
+    ExceptionHandler exceptionHandler = new ExceptionHandler(responseStreamObserver);
+    BiFunction<QueryOuterClass.Response, AtomicLong, MessageHandler<QueryOuterClass.Response, ?>>
+        handlerProducer =
+            (response, inFlight) -> {
+              // decrement inFlight after a delay == 4 * delayBetweenOnCompleteChecks
+              executor.schedule(
+                  inFlight::decrementAndGet,
+                  MessageStreamObserver.DELAY_BETWEEN_ON_COMPLETE_CHECKS_MS * 4,
+                  TimeUnit.MILLISECONDS);
+              return mock(MessageHandler.class);
+            };
+
+    MessageStreamObserver<QueryOuterClass.Response> observer =
+        new MessageStreamObserver<>(
+            responseStreamObserver, exceptionHandler, handlerProducer, executor, 10);
+
+    // when
+    observer.onNext(QueryOuterClass.Response.newBuilder().build());
+
+    // then wait for inFlight = 0 and complete
+    assertDoesNotThrow(
+        observer
+            ::onCompleted); // onCompleted will not return immediately because inFlight is not == 0
```

Comment:
`onCompleted will not return immediately because inFlight is not == 0`

Exactly what I talked about above.. fyi if you apply my changes, you can still verify with:

```
verify(callerStreamObserver, timeout(1000).times(1)).onCompleted();
```